goblet/backends/backend.py

```diff
@@ -0,0 +1,131 @@
+
+    def __init__(self, app, client, func_path, config={}, zip_config=None):
+        self.app = app
+        self.name = app.function_name
+        self.log = logging.getLogger("goblet.backend")
+        self.log.setLevel(logging.INFO)
+        self.zipf = None
+        self.zip_path = get_g_dir() + f"/{self.name}.zip"
+        self.config = GConfig(config=config)
+
+        # specifies which files to be zipped
+        self.zip_config = zip_config or {
+            "include": ["*.py"],
+            "exclude": ["build", "docs", "examples", "test", "tests", "venv"]
+        }
+
+        self.func_path = func_path
+
+        self.client = client
+
+    def _create_zip(self):
+        """Creates initial goblet zipfile"""
+        if not os.path.isdir(get_g_dir()):
+            os.mkdir(get_g_dir())
+        return zipfile.ZipFile(
+            self.zip_path, "w", zipfile.ZIP_DEFLATED
+        )
+
+    def _delta(self, client):
+        """Compares md5 hash between local zipfile and cloudfunction already deployed"""
+        self.zipf.close()
+        with open(self.zip_path, "rb") as fh:
+            local_checksum = base64.b64encode(checksum(fh, hashlib.md5())).decode(
+                "ascii"
+            )
+
+        source_info = client.execute(
+            "generateDownloadUrl", parent_key="name", parent_schema=self.func_path
+        )
+        resp = request("HEAD", source_info["downloadUrl"])
+        deployed_checksum = resp.headers["x-goog-hash"].split(",")[-1].split("=", 1)[-1]
+        modified = deployed_checksum != local_checksum
+        return modified
+
+    def _gcs_upload(self, client, headers, upload_client=None, force=False):
+        self.log.info("zipping source code")
+        self.zipf = self._create_zip()
+        # zip with default include and exclude
+        self._zip()
+        if not force and self.get() and not self._delta(client):
+            self.log.info("No changes detected....")
+            return None
+        self.log.info("uploading source zip to gs......")
+        return self._upload_zip(upload_client or client, headers)
+
+    def _upload_zip(self, client, headers=None) -> dict:
+        """Uploads zipped cloudfunction using generateUploadUrl endpoint"""
+        self.zipf.close()
+        zip_size = os.stat(f".goblet/{self.name}.zip").st_size
+        with open(f".goblet/{self.name}.zip", "rb") as f:
+            resp = client.execute("generateUploadUrl", params={"body": {}})
+            put_headers = {
+                "content-type": "application/zip",
+                "Content-Length": str(zip_size),
+            }
+            if client.version == "v1":
+                put_headers["x-goog-content-length-range"] = "0,104857600"
+
+            requests.put(
+                resp["uploadUrl"],
+                data=f,
+                headers=put_headers,
+            ).raise_for_status()
+
+        self.log.info("function code uploaded")
+
+        return resp
+
+    def get(self):
```

Comment:
some of these are cloudfunction specific so these should go into the v1,v2. we can make a base Cloudfunction class if there is overlap.

Reply:
Ok I like that idea!